src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,469 @@
+                            daiDebt
+                        );
+
+        // Determine the amount of fees to bring back
+        uint256 fees = 0;
+        if (gemBalance > daiDebt) {
+            fees = gemBalance - daiDebt;
+
+            if (_add(amount, fees) > availableLiquidity) {
+                // Don't need safe-math because this is constrained above
+                fees = availableLiquidity - amount;
+            }
+        }
+
+        if (amount == 0 && fees == 0) {
+            emit Unwind(ilk, 0);
+            return;
+        }
+
+        require(amount <= 2 ** 255, "DssDirectDepositHub/overflow");
+
+        // To save gas you can bring the fees back with the unwind
+        uint256 total = _add(amount, fees);
+        join.withdraw(total);
+
+        // normalized debt == erc20 DAI to join (Vat rate for this ilk fixed to 1 RAY)
+
+        address vow = chainlog.getAddress("MCD_VOW");
+        if (mode == Mode.NORMAL) {
+            vat.frob(ilk, address(join), address(join), address(join), -int256(amount), -int256(amount));
+            vat.slip(ilk, address(join), -int256(amount));
+            vat.move(address(join), vow, _mul(fees, RAY));
+        } else if (mode == Mode.MODULE_CULLED) {
+            vat.slip(ilk, address(join), -int256(amount));
+            vat.move(address(join), vow, _mul(total, RAY));
+        } else {
+            // This can be done with the assumption that the price of 1 aDai equals 1 DAI.
+            // That way we know that the prev End.skim call kept its gap[ilk] emptied as the CDP was always collateralized.
+            // Otherwise we couldn't just simply take away the collateral from the End module as the next line will be doing.
+            vat.slip(ilk, end, -int256(amount));
+            vat.move(address(join), vow, _mul(total, RAY));
+        }
+
+        emit Unwind(ilk, amount);
+    }
+
+    function exec(bytes32 ilk_) external {
+        Ilk memory ilk = ilks[ilk_];
+
+        uint256 availableLiquidity = dai.balanceOf(address(ilk.gem));
+
+        if (vat.live() == 0) {
+            // MCD caged
+            require(EndLike(chainlog.getAddress("MCD_END")).debt() == 0, "DssDirectDepositHub/end-debt-already-set");
+            require(ilk.culled == 0, "DssDirectDepositHub/module-has-to-be-unculled-first");
+            _unwind(
+                ilk_,
+                ilk.join,
+                type(uint256).max,
+                availableLiquidity,
+                Mode.MCD_CAGED
+            );
+        } else if (live == 0) {
+            // This module caged
+            _unwind(
+                ilk_,
+                ilk.join,
+                type(uint256).max,
+                availableLiquidity,
+                ilk.culled == 1
+                ? Mode.MODULE_CULLED
+                : Mode.NORMAL
+            );
+        } else {
+            // Normal path
+            (uint256 supplyAmount, uint256 targetSupply) = ilk.join.calcSupplies(availableLiquidity, ilk.bar);
+
```

Comment:
We might want to just have `calcSupplies()` return the just one number (which in Aave's case will be the delta of `targetSupply` and  `supplyAmount`) and an indication of whether to wind or unwind (either boolean or use a signed int). This will help with other type of integrations where the decision is not neccesarily based on this simple substraction. Can work around it by returning one of them as 0 but it will be missleading I think.

Reply:
So if we split out the target supply strategy from the join adapter then the strategy should only provide the desired target DAI amount, and the join adapter should be responsible for reporting what the current debt is and how much can be removed or added. I can't think of any D3M that would fall outside this case.